src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    internal class HostingApplicationDiagnostics
+    {
+        private static readonly double TimestampToTicks = TimeSpan.TicksPerSecond / (double)Stopwatch.Frequency;
+
+        private const string ActivityName = "Microsoft.AspNetCore.Hosting.HttpRequestIn";
+        private const string ActivityStartKey = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start";
+
+        private const string DiagnosticsBeginRequestKey = "Microsoft.AspNetCore.Hosting.BeginRequest";
+        private const string DiagnosticsEndRequestKey = "Microsoft.AspNetCore.Hosting.EndRequest";
+        private const string DiagnosticsUnhandledExceptionKey = "Microsoft.AspNetCore.Hosting.UnhandledException";
+
+        private const string RequestIdHeaderName = "Request-Id";
+        private const string CorrelationContextHeaderName = "Correlation-Context";
+
+
+        private readonly DiagnosticSource _diagnosticSource;
+        private readonly ILogger _logger;
+
+        public HostingApplicationDiagnostics(ILogger logger, DiagnosticSource diagnosticSource)
+        {
+            _logger = logger;
+            _diagnosticSource = diagnosticSource;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context context)
+        {
+            // These enabled checks are virtual dispatch and used twice and so cache to locals
+            var diagnoticsEnabled = _diagnosticSource.IsEnabled(DiagnosticsBeginRequestKey);
+
+            var loggingEnabled = _logger.IsEnabled(LogLevel.Information);
+            var eventLogEnabled = HostingEventSource.Log.IsEnabled();
+
+            if (eventLogEnabled)
+            {
+                context.EventLogEnabled = true;
+                // To keep the hot path short we defer logging in this function to non-inlines
+                RecordRequestStartEventLog(httpContext);
+            }
+
+            // Only make call GetTimestamp if its value will be used, i.e. of the listenters is enabled
+            var startTimestamp = (diagnoticsEnabled || loggingEnabled) ? Stopwatch.GetTimestamp() : 0;
+
+            // Scope may be relevant for a different level of logging, so we always create it
+            // see: https://github.com/aspnet/Hosting/pull/944
+            context.Scope = HostingLoggerExtensions.RequestScope(_logger, httpContext);
+
+            if (loggingEnabled)
+            {
+                // Non-inline
+                LogRequestStarting(httpContext);
```

Comment:
could you log it after activity is started?
It could be quite useful for custom ILogger implementation to use Activity.Current to augment logs